src/core/ready.js

```diff
@@ -49,11 +47,28 @@ jQuery.extend( {
 			return;
 		}
 
-		// If there are functions bound, to execute
-		readyList.resolveWith( document, [ jQuery ] );
+		whenReady = function( fn ) {
+			readyCallbacks.push( fn );
+
+			if ( !++readyFiring ) {
+				while ( readyCallbacks.length ) {
+					fn = readyCallbacks.shift();
+					if ( jQuery.isFunction( fn ) ) {
+						fn.call( document, jQuery );
+					}
+				}
+			}
+
+			readyFiring--;
+		};
+
+		whenReady();
 	}
 } );
 
+// Make jQuery.ready promise compatible (gh-1778)
+jQuery.ready.then = jQuery.fn.ready;
```

Comment:
I think you may want to add some unit tests using a Promise lib, or just plain old native Promises.  Since the implementation of `jQuery.fn.ready` returns `this`, and you're assigning that function to your thenable object, you're essentially getting into this scenario:

```js
var myCrazyPromise = { then: function() { return this; };
Promise.resolve(myCrazyPromise).then(console.log.bind(console));
```

If you run that, you'll see that the `then` handler is never invoked since this is infinitely recursive